org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/PortableURIs.java

```diff
@@ -0,0 +1,300 @@
+		mock.setEType(desc.descriptionEClass);
+		IScope scope = this.globalScopeProvider.getScope(resource, mock, Predicates.<IEObjectDescription>alwaysTrue());
+		Optional<IEObjectDescription> description = Streams.stream(scope.getElements(desc.descriptionQualifiedName)).findFirst();
+		return description.map(d->{
+			EObject container = EcoreUtil.resolve(d.getEObjectOrProxy(), resource);
+			return getEObject(container, desc.descriptionRelativeFragment);
+		}).orElse(null);
+	}
+
+	/**
+	 * Creates and returns a portable URI from the given resource to the targetURI. Returns <code>null</code> is no
+	 * portable URI can be constructed, which is the case if the targetObject is not itself exported or is a child of an
+	 * exported EObject.
+	 * 
+	 * @param sourceResource
+	 *            the resource from which the EObject should later be resolved
+	 * @param targetURI
+	 *            the target URI that should be resolvable by the created portable URI
+	 * 
+	 * @return a portable URI or <code>null</code>
+	 */
+	public URI toPortableURI(StorageAwareResource sourceResource, URI targetURI) {
+		EObject object = sourceResource.getResourceSet().getEObject(targetURI, false);
+		// if it points to some registered ecore, there's no resourceSet and the result is not portable
+		if (object == null || object.eResource().getResourceSet() != null) {
+			return toPortableURI(sourceResource, object);
+		}
+		return null;
+	}
+
+	/**
+	 * Creates and returns a portable URI from the global scope. Returns <code>null</code> is no portable URI can be
+	 * constructed, which is the case if the targetObject is not itself exported or is a child of an exported EObject.
+	 * 
+	 * @param sourceResource
+	 *            the resource from which the EObject should later be resolved
+	 * @param targetObject
+	 *            the target object that should be resolvable by the created portable URI
+	 * 
+	 * @return a portable URI or <code>null</code>
+	 */
+	public URI toPortableURI(StorageAwareResource sourceResource, EObject targetObject) {
+		if (targetObject == null || targetObject.eIsProxy()) {
+			return sourceResource.getURI().appendFragment(StorageAwareResource.UNRESOLVABLE_FRAGMENT);
+		}
+		String portableFragment = getPortableURIFragment(targetObject);
+		if ((portableFragment != null)) {
+			return sourceResource.getURI().appendFragment(portableFragment);
+		}
+		return null;
+	}
+
+	/**
+	 * @return a portable URI fragment, or <code>null</code> if the give EObject isn't itself or is not contained in an
+	 *         exported EObjectDescription
+	 */
+	protected String getPortableURIFragment(EObject obj) {
+		IResourceDescriptions descriptions = resourceDescriptionsProvider.getResourceDescriptions(obj.eResource());
+		IResourceDescription desc = descriptions.getResourceDescription(obj.eResource().getURI());
+		if ((desc == null)) {
+			return null;
+		}
+		return FluentIterable.from(desc.getExportedObjects()).filter(description->{
```

Comment:
Did you purposefully use `FluentIterable` here?

Reply:
In fact I did, but it was trivial to use Streams.stream instead.